### PR TITLE
Pyro 3.16: include in cylc distribution + related improvements

### PR DESCRIPTION
**Pull request by matthewrmshin**
Add Pyro 3.16 under `lib/Pyro/`.

Modify Pyro to memoize `socket.gethost*` calls with a 1 hour expiry.
* Add 1 hour expiry to `cylc.suite_host` as well.

Follow on from #1732.

### CODE REVIEW

**Comment by matthewrmshin** [comment]
This change is done in 3 stages:
1. Add Pyro to cylc distribution.
2. Add memoize functionality to Pyro with similar logic to `cylc.suite_host`.
3. Add expiry logic to the singletons for memoizing.

**Comment by hjoliver** [comment]
Nice.  I guess we can update (simplify) installation documentation in the CUG now.

**Comment by matthewrmshin** [comment]
Comparing master and this branch with `strace -e trace=getsockname -f cylc gscan` suggests that the change is doing the right thing.

**Comment by matthewrmshin** [comment]
(I'll do a `git grep doc/` and see what needs to be updated.)

**Comment by matthewrmshin** [comment]
@hjoliver @benfitzpatrick please review. (It is perhaps easiest to review this change by looking at the individual changesets.)

**Comment by hjoliver** [comment]
Review 1: looks good, test battery passes.  Tested with a short EXPIRE time.  Over to you Ben...

**Comment by matthewrmshin** [comment]
(Branch re-based against master.)

File: doc/cug.tex [hunk 1]
```diff
@@ -445,6 +441,12 @@ \section{Required Software}
     \item ImageMagick (for image scaling)
 \end{myitemize}
 
+Note: Older versions of cylc requires Pyro-3 (Python Remote Objects). Current
```
Comment (benfitzpatrick) [inline doc/cug.tex]: require

**Comment by benfitzpatrick** [comment]
One comment, otherwise looks fine to me.

File: doc/cug.tex (hunk 1 above)
Comment (matthewrmshin) [inline doc/cug.tex]: Fixed.

**Comment by matthewrmshin** [comment]
Typo fixed.

**Pull request merged**